**Hyper parameter tuning**

- The parameters or arguments of model can change by user to get the good model performance.

- Grid search

- Random search

- Assume that in Decsision tree we have following hyper parameter

    - criterion : {"gini", "entropy", "log_loss"}
        
    - splitter : {"best", "random"}
     
    - max depth : {1,2,3,4,5,6}
        
- By default we have criterion 'gini' is there

- But the model might be performed when criterion is 'entropy' also

- we need to test the combinations by providing different values

- To try the combinations we have two methods

    - Grid search:
        
        - it is a step by step by consider the all combinations
        
        - in criterion: 3,splitter: 2, max depth=6
                
        - The total combinations are : 3x2x6=36
            
    - Random search:
        
        - It will not do all 36 possibile combiantions
        
        - It will select randomly some combinations
        
- Generally we do Grid search 

- By doing this we have term called Cross validation


**Cross validation**:
    
- Cross validation: CV
    
- CV=4 means we devide data randomly into 4 parts

- Out of these 4 parts every time 3 parts consider as train data, 1 parts consider as test data

- P1  P2  P3  P4  are 4 parts of data

- Combination1:  P1 + P2 + P3 are train data  and P4 as test data
    
    - For Combination-1 will try to imporove model performance by doing Hyper parameter tuning
    
    - Assume there 36 comninations are available in DT
    
    - For the Combination1 will do all 36
 
    - so we have 36 hyper parameter comninations for each combination will have one accuracy
 
    - so we have total 36 accuracy values

- Combination2:  P1 + P2 + P4   are train data  and   P3 as test data
    
    - For Combination-2 will try to imporove model performance by doing Hyper parameter tuning
    
    - Assume there 36 comninations are available in DT
    
    - For the Combination-2 will do all 36 

- Combination3:  P1 + P3 + P4  are train data and P2 as test data
    
    - For Combination-3 will try to imporove model performance by doing Hyper parameter tuning
    
    - Assume there 36 comninations are available in DT
    
    - For the Combination-3 will do all 36 

- Combination-4: P2 + P3 + P4  are train data and P1 as test data
    
    - For Combination-4 will try to imporove model performance by doing Hyper parameter tuning
    
    - Assume there 36 comninations are available in DT
    
    - For the Combination-4 will do all 36
    
- Every combination gives 36 accurcy values corresponding each hyper parameter

- will calculate all the average accuracy of combinations individual accuracy w.r.t to each hyper parameter

- so that will get 36 average accuracy values w.r.t each hyper parameter

- Out of these 36 the maximum one is our best accuracy and corresponding hyper parameters are best parameters  

In [5]:
import pandas as pd
indexes=[f"h_{i}" for i in range(1,37)]
col1=[f"combination_1_acc_{i}" for i in range(1,37)]
col2=[f"combination_2_acc_{i}" for i in range(1,37)]
col3=[f"combination_3_acc_{i}" for i in range(1,37)]
col4=[f"combination_4_acc_{i}" for i in range(1,37)]
col5=[f"avg_acc_{i}" for i in range(1,37)]
cols=['C1','C2','C3','C4','Avg acc']
pd.DataFrame(zip(col1,col2,col3,col4,col5),index=indexes,
            columns=cols)

C1                    C2                    C3  \
h_1    combination_1_acc_1   combination_2_acc_1   combination_3_acc_1   
h_2    combination_1_acc_2   combination_2_acc_2   combination_3_acc_2   
h_3    combination_1_acc_3   combination_2_acc_3   combination_3_acc_3   
h_4    combination_1_acc_4   combination_2_acc_4   combination_3_acc_4   
h_5    combination_1_acc_5   combination_2_acc_5   combination_3_acc_5   
h_6    combination_1_acc_6   combination_2_acc_6   combination_3_acc_6   
h_7    combination_1_acc_7   combination_2_acc_7   combination_3_acc_7   
h_8    combination_1_acc_8   combination_2_acc_8   combination_3_acc_8   
h_9    combination_1_acc_9   combination_2_acc_9   combination_3_acc_9   
h_10  combination_1_acc_10  combination_2_acc_10  combination_3_acc_10   
h_11  combination_1_acc_11  combination_2_acc_11  combination_3_acc_11   
h_12  combination_1_acc_12  combination_2_acc_12  combination_3_acc_12   
h_13  combination_1_acc_13  combination_2_acc_13  combination_3_acc_13   
h_14  combination_1_acc_14  combination_2_acc_14  combination_3_acc_14   
h_15  combination_1_acc_15  combination_2_acc_15  combination_3_acc_15   
h_16  combination_1_acc_16  combination_2_acc_16  combination_3_acc_16   
h_17  combination_1_acc_17  combination_2_acc_17  combination_3_acc_17   
h_18  combination_1_acc_18  combination_2_acc_18  combination_3_acc_18   
h_19  combination_1_acc_19  combination_2_acc_19  combination_3_acc_19   
h_20  combination_1_acc_20  combination_2_acc_20  combination_3_acc_20   
h_21  combination_1_acc_21  combination_2_acc_21  combination_3_acc_21   
h_22  combination_1_acc_22  combination_2_acc_22  combination_3_acc_22   
h_23  combination_1_acc_23  combination_2_acc_23  combination_3_acc_23   
h_24  combination_1_acc_24  combination_2_acc_24  combination_3_acc_24   
h_25  combination_1_acc_25  combination_2_acc_25  combination_3_acc_25   
h_26  combination_1_acc_26  combination_2_acc_26  combination_3_acc_26   
h_27  combination_1_acc_27  combination_2_acc_27  combination_3_acc_27   
h_28  combination_1_acc_28  combination_2_acc_28  combination_3_acc_28   
h_29  combination_1_acc_29  combination_2_acc_29  combination_3_acc_29   
h_30  combination_1_acc_30  combination_2_acc_30  combination_3_acc_30   
h_31  combination_1_acc_31  combination_2_acc_31  combination_3_acc_31   
h_32  combination_1_acc_32  combination_2_acc_32  combination_3_acc_32   
h_33  combination_1_acc_33  combination_2_acc_33  combination_3_acc_33   
h_34  combination_1_acc_34  combination_2_acc_34  combination_3_acc_34   
h_35  combination_1_acc_35  combination_2_acc_35  combination_3_acc_35   
h_36  combination_1_acc_36  combination_2_acc_36  combination_3_acc_36   

                        C4     Avg acc  
h_1    combination_4_acc_1   avg_acc_1  
h_2    combination_4_acc_2   avg_acc_2  
h_3    combination_4_acc_3   avg_acc_3  
h_4    combination_4_acc_4   avg_acc_4  
h_5    combination_4_acc_5   avg_acc_5  
h_6    combination_4_acc_6   avg_acc_6  
h_7    combination_4_acc_7   avg_acc_7  
h_8    combination_4_acc_8   avg_acc_8  
h_9    combination_4_acc_9   avg_acc_9  
h_10  combination_4_acc_10  avg_acc_10  
h_11  combination_4_acc_11  avg_acc_11  
h_12  combination_4_acc_12  avg_acc_12  
h_13  combination_4_acc_13  avg_acc_13  
h_14  combination_4_acc_14  avg_acc_14  
h_15  combination_4_acc_15  avg_acc_15  
h_16  combination_4_acc_16  avg_acc_16  
h_17  combination_4_acc_17  avg_acc_17  
h_18  combination_4_acc_18  avg_acc_18  
h_19  combination_4_acc_19  avg_acc_19  
h_20  combination_4_acc_20  avg_acc_20  
h_21  combination_4_acc_21  avg_acc_21  
h_22  combination_4_acc_22  avg_acc_22  
h_23  combination_4_acc_23  avg_acc_23  
h_24  combination_4_acc_24  avg_acc_24  
h_25  combination_4_acc_25  avg_acc_25  
h_26  combination_4_acc_26  avg_acc_26  
h_27  combination_4_acc_27  avg_acc_27  
h_28  combination_4_acc_28  avg_acc_28  
h_29  combination_4_acc_29  avg_acc_29  
h_30  combination_4_acc_30  avg_acc_30  
h_31  combinat

$Step-1$

**Import the packages**

In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme(color_codes=True)             # To get diffent different colors
pd.set_option('display.max_columns', None)  # To display the max columns
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,\
                            precision_score,\
                            recall_score,\
                            f1_score,\
                            classification_report,\
                            roc_auc_score,roc_curve,auc
from sklearn.metrics import confusion_matrix,ConfusionMatrixDisplay

$Step-2$:
    
**Read the data**

In [7]:
data=pd.read_csv("preprocessd_data.csv")
data

gender       age  no_of_days_subscribed  multi_screen  mail_subscribed  \
0          0 -0.264643              -0.946659             0                0   
1          0  0.029654               1.235009             0                0   
2          0  2.580235               0.658246             0                0   
3          0 -1.441834               0.783629             0                1   
4          0  0.127754               2.288228             0                0   
...      ...       ...                    ...           ...              ...   
3401       1 -0.805656               0.479710             0                0   
3402       0  0.477826               0.674425             0                0   
3403       0 -0.550521               0.426099             0                1   
3404       0  0.004651              -0.444637             0                0   
3405       0 -0.057028               0.317748             0                0   

      weekly_mins_watched  minimum_daily_mins  weekly_max_night_mins  \
0               -1.508785            0.722258              -0.944404   
1                0.304821           -0.901748              -0.688523   
2               -2.266627            0.613991              -0.483818   
3                0.638123           -0.252146               0.079121   
4               -0.333851            0.253101              -0.893228   
...                   ...                 ...                    ...   
3401             0.887802            0.552184               0.865765   
3402             2.038061           -0.931071               0.567257   
3403            -0.410669            0.344847              -0.729485   
3404             1.359751           -0.847121              -0.715010   
3405             2.844852            0.098115              -0.550251   

      videos_watched  customer_support_calls  churn  
0          -1.403000               -0.415473      0  
1          -0.599124                0.347522      0  
2          -1.403000                2.636507      1  
3          -0.197185                1.110517      0  
4           1.008630               -0.415473      0  
...              ...                     ...    ...  
3401       -0.495634               -0.415473      1  
3402        2.100185                0.235227      1  
3403       -0.403904                1.873512      1  
3404       -0.197185               -1.178468      1  
3405        0.606692               -1.178468      1  

[3406 rows x 11 columns]

$Step-3$

**Divide data into input and output**

In [8]:
# Here the target column is Customer Status
X = data.drop('churn', axis=1)
y = data['churn']

$Step-4$:
    
**Train test split**

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2,random_state=0)
print(X_train.shape),
print(X_test.shape),
print(y_train.shape),
print(y_test.shape)

(2724, 10)
(682, 10)
(2724,)
(682,)


$Step-5$

**Read the base model**

In [10]:
from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.tree import DecisionTreeClassifier
grid_tree=DecisionTreeClassifier()   # Base model
grid_tree

# we are not fitting the data

DecisionTreeClassifier()

$Step-6$

**Create the parameter file**

In [11]:
grid_tree.get_params()

{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'random_state': None,
 'splitter': 'best'}

**max_depth**:
    
    - The maximum depth of the tree. If None, then nodes are expanded until all leaves are pure or until all leaves contain less        than  min_samples_split samples.
    
**min_samples_split**:
     
      - The minimum number of samples required to split an internal node:
      
**min_samples_leaf**:
    
       - The minimum number of samples required to be at a leaf node.

In [12]:
# You need to create dictonary with hyper parameters
param_grid = {
    "criterion":['gini','entropy'], # 2
    'max_depth': [3, 4, 5, 6, 7, 8], # 6
    'min_samples_split': [2, 3, 4],  # 3
    'min_samples_leaf': [1, 2, 3, 4], # 4
    'random_state': [0, 42] # 2
}

# 2*6*3*4*2=288

$Step-7$

**Apply Grid Search CV**

In [13]:
# now these parameters to apply in a DT model

grid_search = GridSearchCV(grid_tree,  # Base model
                           param_grid, # params 
                           scoring='accuracy', # metric
                           cv=5,
                           verbose=True)

# Verbose=True I will show in Deep learning indetail

In [14]:
grid_search

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 4, 5, 6, 7, 8],
                         'min_samples_leaf': [1, 2, 3, 4],
                         'min_samples_split': [2, 3, 4],
                         'random_state': [0, 42]},
             scoring='accuracy', verbose=True)

In [15]:
dir(grid_search)

['__abstractmethods__',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__sklearn_clone__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_build_request_for_signature',
 '_check_feature_names',
 '_check_n_features',
 '_check_refit_for_multimetric',
 '_doc_link_module',
 '_doc_link_template',
 '_doc_link_url_param_generator',
 '_estimator_type',
 '_format_results',
 '_get_default_requests',
 '_get_doc_link',
 '_get_metadata_request',
 '_get_param_names',
 '_get_routed_params_for_fit',
 '_get_scorers',
 '_get_tags',
 '_more_tags',
 '_parameter_constraints',
 '_repr_html_',
 '_repr_html_inner',
 '_repr_mimebundle_',
 '_required_parameters',
 '_run

$Step-8$

**Fit the data**

In [16]:
import time
start=time.time()
grid_search.fit(X_train,y_train)  
# Generally we will pass the entire data
# X_train= input data
# y_train= output data
end=time.time()
print("Total time taken is:",(end-start))

Fitting 5 folds for each of 288 candidates, totalling 1440 fits
Total time taken is: 13.317485570907593


In [17]:
dir(grid_search)

['__abstractmethods__',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__sklearn_clone__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_build_request_for_signature',
 '_check_feature_names',
 '_check_n_features',
 '_check_refit_for_multimetric',
 '_doc_link_module',
 '_doc_link_template',
 '_doc_link_url_param_generator',
 '_estimator_type',
 '_format_results',
 '_get_default_requests',
 '_get_doc_link',
 '_get_metadata_request',
 '_get_param_names',
 '_get_routed_params_for_fit',
 '_get_scorers',
 '_get_tags',
 '_more_tags',
 '_parameter_constraints',
 '_repr_html_',
 '_repr_html_inner',
 '_repr_mimebundle_',
 '_required_parameters',
 '_run

$Step-9$:
    
**Get the best parameters**

In [18]:
grid_search.best_estimator_

# Estimator means model : 144 models

# best estimator is your model
# if you use this model 
# then we need to find all the metrics

DecisionTreeClassifier(criterion='entropy', max_depth=8, min_samples_leaf=4,
                       random_state=0)

In [19]:
grid_search.best_score_
# not a single accuracy
# avg accuracy of all 5 parts

0.8649001618996222

In [20]:
grid_search.best_params_

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

- Estimator means best Decision Tree model

- best params means the parameters which are used in best model

In [21]:
grid_search.cv_results_

{'mean_fit_time': array([0.00478926, 0.00484715, 0.00431333, 0.00417204, 0.00435491,
        0.00518928, 0.0052412 , 0.00494218, 0.00457444, 0.00460067,
        0.00454187, 0.00447054, 0.00414753, 0.00493903, 0.00534   ,
        0.00521803, 0.00490303, 0.00511498, 0.00515285, 0.00492654,
        0.00473232, 0.00487614, 0.00481105, 0.00436716, 0.00571666,
        0.00644932, 0.00576267, 0.00554099, 0.00600581, 0.00563745,
        0.00532217, 0.00566998, 0.00567622, 0.00613279, 0.00560842,
        0.00594549, 0.00572829, 0.00608311, 0.00577054, 0.0056047 ,
        0.00590935, 0.00548787, 0.00541663, 0.00540309, 0.00545597,
        0.00602784, 0.0055542 , 0.00564628, 0.00605998, 0.00607891,
        0.00637693, 0.00664849, 0.00606303, 0.00670147, 0.00593982,
        0.00565643, 0.0064446 , 0.01271038, 0.00716219, 0.00612659,
        0.00723681, 0.00680189, 0.00624804, 0.00551353, 0.0055531 ,
        0.00597816, 0.00585194, 0.00571823, 0.00604243, 0.00619693,
        0.0065948 , 0.00625982,

In [22]:
grid_search.cv_results_.keys()
#grid_search.cv_results_['params']

dict_keys(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time', 'param_criterion', 'param_max_depth', 'param_min_samples_leaf', 'param_min_samples_split', 'param_random_state', 'params', 'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score', 'mean_test_score', 'std_test_score', 'rank_test_score'])

In [23]:
#len(grid_search.cv_results_['split0_test_score'])
c1=grid_search.cv_results_['split0_test_score']
c2=grid_search.cv_results_['split1_test_score']
c3=grid_search.cv_results_['split2_test_score']
c4=grid_search.cv_results_['split3_test_score']
c5=grid_search.cv_results_['split4_test_score']
mean_score=grid_search.cv_results_['mean_test_score']
df=pd.DataFrame(zip(c1,c2,c3,c4,c5,mean_score),
             index=grid_search.cv_results_['params'],
            columns=['C1','C2','C3','C4','C5','mean_test_score'])
avg=[sum(i[:-1])/5 for i in df.values]
max(avg)
df['Average']=avg
df

C1        C2  \
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.827523  0.829358   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.827523  0.829358   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.827523  0.829358   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.827523  0.829358   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.827523  0.829358   
...                                                      ...       ...   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.877064  0.871560   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.880734  0.880734   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.877064  0.871560   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.880734  0.880734   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.877064  0.871560   

                                                          C3        C4  \
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.814679  0.825688   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.814679  0.825688   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.814679  0.825688   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.814679  0.825688   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.814679  0.825688   
...                                                      ...       ...   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.851376  0.858716   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.849541  0.860550   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.851376  0.858716   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.849541  0.860550   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.851376  0.858716   

                                                          C5  mean_test_score  \
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.810662         0.821582   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.810662         0.821582   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.810662         0.821582   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.810662         0.821582   
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.810662         0.821582   
...                                                      ...              ...   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.845588         0.860861   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.852941         0.864900   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.845588         0.860861   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.852941         0.864900   
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.845588         0.860861   

                                                     Average  
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.821582  
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.821582  
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.821582  
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.821582  
{'criterion': 'gini', 'max_depth': 3, 'min_samp...  0.821582  
...                                                      ...  
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.860861  
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.864900  
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.860861  
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.864900  
{'criterion': 'entropy', 'max_depth': 8, 'min_s...  0.860861  

[288 rows x 7 columns]

In [24]:
df['Average'].max(),grid_search.best_score_

(0.8649001618996222, 0.8649001618996222)

In [25]:
df.values[0][:-1]

array([0.82752294, 0.8293578 , 0.8146789 , 0.82568807, 0.81066176,
       0.82158189])

In [26]:
# Best Parameter
df[['Average']].idxmax().values[0]

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

In [27]:
id=df[['Average']].idxmax()
id.values

array([{'criterion': 'entropy', 'max_depth': 8, 'min_samples_leaf': 4, 'min_samples_split': 2, 'random_state': 0}],
      dtype=object)

In [28]:
# df=pd.DataFrame(zip([1,2,3],['A','B','C']),
#                columns=['Col1','Col2'])
# df['Col1'].idxmax() # index=2
# df.take([2])
# df.loc[2]

In [ ]:
# Best accuracy
df[['Average']].max().values[0]

In [ ]:
id=df['Average'].idxmax()
l=tuple(id.items())
l
df.loc[l]

**Cross Validation score**

- Now we use the best model

- Perform the Cross validation score

- Best model means it takes the best parameter

- Cross validation 5 means , it gives 5 accuracy values

- Take the average of all accuracy, again that should be equal to besta accuracy

In [ ]:
# Assume that we already know the best model
# best paramters with out doing Grid Search
# Now I want to apply Cross valiation
# In the cross validation every part involved in training and testing
# cross_val_score cv=5 by default

In [ ]:
accuracy_list=cross_val_score(grid_search.best_estimator_,
                X_train,
                y_train,
                cv=5,
                scoring='accuracy')
accuracy_list

# It will give each part accuracy
# by using best model (best estimator)

In [ ]:
accuracy_list.mean()

In [ ]:
## All together

#================================ Step-1: Get the base model===========================================
from sklearn.model_selection import GidSearrchCV,cross_val_score
grid_tree=DecisionTreeClassifier()   # Base model
DT_params=[keys for keys in grid_tree.get_params()]
print("avialable parameters for model:",DT_params)

print("==========================================================================================")

#================================ Step-2: Create a dictionary for params=================================
param_grid = {
    "criterion":['gini','entropy'],
    'max_depth': [3, 4, 5, 6, 7, 8],
    'min_samples_split': [2, 3, 4],
    'min_samples_leaf': [1, 2, 3, 4],
    'random_state': [0, 42]
}

print("Processing parameters are:",param_grid)

print("===============================================================================================")

#============================== Step-3: Apply the Grid serach======================================
grid_search = GridSearchCV(grid_tree,  # Base modl
                           param_grid, # params 
                           scoring='accuracy', # metric
                           cv=5,
                          verbose=True)

#============================= Step-4: fit the model with train data ===============================
grid_search.fit(X_train,y_train)

#============================= Step-5: Get the best parameters=====================================
best_params=grid_search.best_params_
best_score=grid_search.best_score_
print("best params:",best_params)
print("best train accuracy:",best_score)

**Identify the predictions and Metrics by using best parameters**

- Out of 1440 fittings, one best fit model is outcome

- Best fit means it will give the best parameters

- For me the best parameter combination is

    - entropy
    
    - depth=6
    
    - sample leaf=3
    
    - sampl split=2
    
    - random state=0
    
- Our goal is by using above parameter we need to find the predictions

- And will evaluate model performance

- For that Import Decision Tree and pass the best parameters

- Now fit that model with X_train, y_train

- Get the predictions by passing X_test

- Compare the predictions with y_test

- Calculate

    - Accuracy
    
    - Precision
    
    - Recall
    
    - F1 score
    
    - Roc-Auc curve

In [ ]:
grid_search.best_params_
grid_search.best_estimator_

In [ ]:
from sklearn.tree import DecisionTreeClassifier
# dtree=DecisionTreeClassifier(criterion='entropy',
#                             max_depth=8,
#                             min_samples_leaf=4,
#                             min_samples_split=2,
#                             random_state=0)
# dtree.fit(X_train, y_train)

# # =============================Step-5:  Predictions============================================================

# y_pred_dt=dtree.predict(X_test)
best_model=grid_search.best_estimator_
y_pred_dt=best_model.predict(X_test)
# ============================ Step-6: Metrics==================================================================

acc_dt= round(accuracy_score(y_test,y_pred_dt)*100,2)
f1_dt=round(f1_score(y_test,y_pred_dt),2)
precision_dt=round(precision_score(y_test,y_pred_dt),2)
recall_dt=round(recall_score(y_test,y_pred_dt),2)

print("accuray is:",acc_dt)
print("F1 is:",f1_dt)
print("Precision is:",precision_dt)
print("Recall is:",recall_dt)
print(classification_report(y_test,y_pred_dt))

# ================================Step-7:Confusion matrix=========================================================================

cmt=confusion_matrix(y_test,y_pred_dt)

disp=ConfusionMatrixDisplay(confusion_matrix=cmt,
                            display_labels = [False, True])
disp.plot()
plt.grid(False)
plt.show()


tn, fp, fn, tp = confusion_matrix(y_test,y_pred_dt).ravel()
print("True negative:",tn)
print("False postive:",fp)
print("False negative:",fn)
print("True postive:",tp)

#=======================================Step-8: ROC-AUC curve================================================================

y_dt_pred_prob=best_model.predict_proba(X_test)[:,1]   # Class-1 probabilities
fpr,tpr,threshold=roc_curve(y_test,y_dt_pred_prob) 
plt.plot([0,1],[0,1],color="navy",lw=2,label="Random-Model")
plt.plot(fpr,tpr,color="darkorange",lw=2, label="Decision-Tree Model")
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.title("Receiver operating characteristic :ROC-AUC")
plt.legend()
plt.show()
print("Computed Area Under the Curve (AUC)",auc(fpr, tpr))

In [ ]:
# Decision tree will provide imprtant features also
# Information gain values 
dtree.feature_importances_

In [ ]:
imp_df = pd.DataFrame({
    "Feature Name": X_train.columns,
    "Importance": dtree.feature_importances_
})
fi = imp_df.sort_values(by="Importance", ascending=False)
fi

# This concept is valid only for DT
# you can use these features top 10 again develop the model
# Dont use these feature in another model